### PR TITLE
implementing proper parallel processing with better memory management

### PR DESCRIPTION
**Pull request by salt431**
my fork has been tested on an i5-4570 in legend of zelda breath of the wild, and I observed a ~17% performance improvement implementing proper parallel processing in the address space manager of the emulated CPU, along with fixes to the memory management of the memory helper.

### CODE REVIEW

**Comment by salt431** [comment]
you may download and test a compiled windows version of the build here: https://github.com/salt431/Ryujinx/releases/tag/1.0

**Comment by riperiperi** [comment]
Do not link compiled versions of builds as PR comments. We have our own CI that builds the changes.

File: src/Ryujinx.Cpu/AppleHv/HvAddressSpace.cs [hunk 1]
```diff
@@ -41,27 +43,27 @@ public HvAddressSpace(MemoryBlock backingMemory, ulong asSize)
...
-                // 0x0: Synchronous
-                // 0x80: IRQ
-                // 0x100: FIQ
-                // 0x180: SError
-                _kernelCodeBlock.Write(KernelRegionCodeOffset + offset, 0xD41FFFE2u); // HVC #0xFFFF
-                _kernelCodeBlock.Write(KernelRegionCodeOffset + offset + 4, 0xD69F03E0u); // ERET
-            }
+{
+    // Write exception handlers.
+    Parallel.For((int)0UL, (int)0x800UL, (offset, state) =>
```
Comment (riperiperi) [inline src/Ryujinx.Cpu/AppleHv/HvAddressSpace.cs:48]: This is completely unnecessary, the code is run once and writing two int32 values is not enough to justify the setup/teardown of parallel tasks.

File: src/Ryujinx.Cpu/MemoryHelper.cs [hunk 2]
```diff
@@ -28,21 +30,20 @@ public static void FillWithZeros(IVirtualMemoryManager memory, ulong position, i
 
         public unsafe static T Read<T>(IVirtualMemoryManager memory, ulong position) where T : unmanaged
         {
-            return MemoryMarshal.Cast<byte, T>(memory.GetSpan(position, Unsafe.SizeOf<T>()))[0];
+            ReadOnlySpan<byte> span = memory.GetSpan(position, Unsafe.SizeOf<T>());
+            return MemoryMarshal.Read<T>(span);
```
Comment (riperiperi) [inline src/Ryujinx.Cpu/MemoryHelper.cs:34]: Can you explain what this is meant to be doing? It just seems to be jumbing around MemoryMarshal methods in a way that roughly does the same thing.

File: src/Ryujinx.Cpu/MemoryHelper.cs [hunk 3]
```diff
@@ -28,21 +30,20 @@ public static void FillWithZeros(IVirtualMemoryManager memory, ulong position, i
...
         }
 
         public unsafe static ulong Write<T>(IVirtualMemoryManager memory, ulong position, T value) where T : unmanaged
         {
-            ReadOnlySpan<byte> data = MemoryMarshal.Cast<T, byte>(MemoryMarshal.CreateReadOnlySpan(ref value, 1));
-
-            memory.Write(position, data);
-
-            return (ulong)data.Length;
+            Span<byte> span = MemoryMarshal.AsBytes(MemoryMarshal.CreateSpan(ref value, 1));
```
Comment (riperiperi) [inline src/Ryujinx.Cpu/MemoryHelper.cs:39]: Why no longer read-only? That's what it is.

File: src/Ryujinx.Cpu/MemoryHelper.cs [hunk 4]
```diff
@@ -1,16 +1,18 @@
-﻿using Microsoft.IO;
-using Ryujinx.Common.Memory;
-using Ryujinx.Memory;
 using System;
 using System.IO;
 using System.Runtime.CompilerServices;
 using System.Runtime.InteropServices;
 using System.Text;
+using Microsoft.IO;
```
Comment (riperiperi) [inline src/Ryujinx.Cpu/MemoryHelper.cs:6]: These are no longer sorted.

File: src/Ryujinx.Cpu/AppleHv/HvAddressSpace.cs (hunk 1 above)
Comment (salt431) [reply src/Ryujinx.Cpu/AppleHv/HvAddressSpace.cs:48]: for me on my 10-year old quad-core CPU this improved performance in breath of the wild by around 17% as the 32-bit engine had a benefit from the parallel tasks

File: src/Ryujinx.Cpu/MemoryHelper.cs (hunk 2 above)
Comment (salt431) [reply src/Ryujinx.Cpu/MemoryHelper.cs:34]: this was just to improve readability for myself

File: src/Ryujinx.Cpu/AppleHv/HvAddressSpace.cs (hunk 1 above)
Comment (riperiperi) [reply src/Ryujinx.Cpu/AppleHv/HvAddressSpace.cs:48]: This code is not executed on your 10-year old quad core i5 CPU, it is only run on Apple Silicon processors. In macs.

File: src/Ryujinx.Cpu/MemoryHelper.cs (hunk 3 above)
Comment (salt431) [reply src/Ryujinx.Cpu/MemoryHelper.cs:39]: this improved frametime stability in super smash brothers ultimate. especially when loading in new maps that have not been cached in yet.

File: src/Ryujinx.Cpu/AppleHv/HvAddressSpace.cs (hunk 1 above)
Comment (amayra) [reply src/Ryujinx.Cpu/AppleHv/HvAddressSpace.cs:48]: so my i7 3770 will benefit from this too?

Comment (jduncanator) [reply src/Ryujinx.Cpu/AppleHv/HvAddressSpace.cs:48]: No. This code is only used on ARM-based Apple products (M1/M2)

**Comment by AcK77** [comment]
As riperiperi said, this doesnt do anything. At least it doesn't do what it pretend.
I'm closing this.

**Pull request closed without merging**